blockmanager.go

```diff
@@ -347,21 +357,100 @@ func (b *blockManager) cfHandler() {
 					time.Sleep(QueryTimeout)
 				}
 			}
-			b.getAllCFHeaders(goodCheckpoints, store, fType)
-			mtxChkpt.Lock()
-			checkpoints[fType] = goodCheckpoints
-			mtxChkpt.Unlock()
-			wg.Done()
-		}(fType, store)
+
+			// Get all the headers up to the last known good
+			// checkpoint.
+			b.getCheckpointedCFHeaders(goodCheckpoints, store,
+				fType)
+
+			// Catch the filter headers up to the tip of the
+			// block header store and then stay caught up with
+			// each signal while watching for reorgs. We ignore
+			// checkpoints because at this point, we're past
+			// any valid checkpoints.
+			ticker := time.NewTicker(QueryTimeout)
+			defer ticker.Stop()
+			for {
+				// Get/write the next set of cfheaders, if any.
+				err = b.getUncheckpointedCFHeaders(store, fType)
+				if err != nil {
+					log.Debugf("couldn't get "+
+						"uncheckpointed headers for "+
+						"%v: %v", fType, err)
+				}
+
+				// Quit if requested.
+				select {
+				case <-b.quit:
+					return
+				case <-ticker.C:
+				}
+
+			}
+		}(fType, storeLookup)
 	}
 	wg.Wait()
+}
+
+// getUncheckpointedCFHeaders gets the next batch of cfheaders from the
+// network, if it can, and resolves any conflicts between them. It then writes
+// any verified headers to the store.
+func (b *blockManager) getUncheckpointedCFHeaders(
+	store *headerfs.FilterHeaderStore, fType wire.FilterType) error {
+
+	// Get the filter header store's chain tip.
+	_, filtHeight, err := store.ChainTip()
+	if err != nil {
+		return fmt.Errorf("error getting filter chain tip: %v", err)
+	}
+	startHeight := filtHeight + 1
 
+	_, blockHeight, err := b.server.BlockHeaders.ChainTip()
+	if err != nil {
+		return fmt.Errorf("error getting block chain tip: %v", err)
+	}
+	if blockHeight < filtHeight {
+		return fmt.Errorf("reorg in progress, waiting to get "+
+			"uncheckpointed cfheaders (block height %d, filter "+
+			"height %d", blockHeight, filtHeight)
+	}
+	if blockHeight == filtHeight {
+		return fmt.Errorf("cfheaders already caught up to blocks")
+	}
+
+	// Query all peers for the responses.
+	headers := b.getCFHeadersForAllPeers(startHeight, fType)
+	if len(headers) == 0 {
+		return fmt.Errorf("couldn't get cfheaders from peers")
+	}
+
+	// For each header, go through and check whether all headers messages
```

Comment:
Why can't we just go checkpoints all the way? Or is the idea that this will only be a few blocks at most that were added after we got the full set of checkpoints, but before this?